### PR TITLE
Typescript refactor

### LINKED ISSUE #45: Keyboard shortcut request
It would be nice to have keyboard shortcuts to move between subtasks

### PR DESCRIPTION
**Pull request by csolbs24**
# Typescript Refactor

## Description

- Massive update with lots of under the hood changes.
- Added a toolbox item for changing annotation color.
- Added a toolbox item for changing annotation size.
- Added a toolbox item for filtering out annotations based on annotation confidence.
- Added support for adding diffrent Submit buttons, such as a Reject button or Failure button.
- Added a display for annotation's confidence on hover.
- Added keybinds for most operations.

Also Closes #45 

## PR Checklist

- [x] Merged latest main
- [x] Version number in `package.json` has been bumped since last release
- [x] Version numbers match between package `package.json` and `src/version.js`
- [x] Ran `npm install` and `npm run build` AFTER bumping the version number
- [x] Updated documentation if necessary (currently just in `api_spec.md`)
- [x] Added changes to `changelog.md`

## Breaking API Changes

This very well may break something. This pr contains all of the changes from the past ~8 months where we went added a lot of features and refactored much of the code.

### CODE REVIEW

File: src/drawing_utilities.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+    "tan":"#d2b48c","teal":"#008080","thistle":"#d8bfd8","tomato":"#ff6347","turquoise":"#40e0d0",
+
+    "violet":"#ee82ee",
+
+    "wheat":"#f5deb3","white":"#ffffff","whitesmoke":"#f5f5f5",
+
+    "yellow":"#ffff00","yellowgreen":"#9acd32"};
+
+    if (typeof colors[color.toLowerCase()] != 'undefined'){
```
Comment (TrevorBurgoyne) [inline src/drawing_utilities.ts:144]:
```suggestion
    if (color.toLowerCase() in colors){
```

File: src/drawing_utilities.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    if (final_hex.length == 7) {
+        return final_hex
+    } else {
+        return base_color_hex
+    }
+}
+
+/*takes in a string of any valid css color and returns its hex value
+if given string is not a valid css color, returns the string passed in */
+export function color_to_hex(color: string) {
```
Comment (TrevorBurgoyne) [inline src/drawing_utilities.ts:87]: is there no cleaner way to include css `key: color` pairs? lol

File: src/html_builder.ts [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+            toolbox_html += `
+            </div>`;
+
+            // Add dialog to the document
+            // front_subtask_dialog_container_jq.append(dialog_html);
+            // $("#" + ul.subtasks[st]["idd_id"]).attr("id", ul.subtasks[st]["idd_id_front"]);
+            front_subtask_dialog_container_jq.append(front_dialog_html_v2); // TODO(new3d) MOVE THIS TO GLOB BOX -- superimpose atop thee anchor already there when needed, no remove and add back
+            subtask_dialog_container_jq.append(dialog_html_v2);
+            // console.log(dialog_html);
+            // console.log(dialog_html_v2);
```
Comment (TrevorBurgoyne) [inline src/html_builder.ts:369]: clean up comments

File: src/index.js [hunk 4]
```diff
@@ -1081,34 +433,34 @@ export class ULabel {
...
+        //         const save_success = await ul.config["done_callback"].bind(ul)(submit_payload);
+        //         ul.set_saved(!(save_success === false));
+        //     }
+        //     catch (err) {
+        //         console.log("Error waiting for submit script.")
+        //         console.log(err);
+        //         ul.set_saved(false);
+        //     }
+        // });
+
```
Comment (TrevorBurgoyne) [inline src/index.js:462]:
```suggestion
```
Remove if not needed

File: src/index.js [hunk 5]
```diff
@@ -1404,27 +720,28 @@ export class ULabel {
...
-                "display_name": raw_subtask["display_name"] || subtask_key,
-                "read_only": ("read_only" in raw_subtask) && (raw_subtask["read_only"] === true),
-                "inactive_opacity": 0.4
-            };
+            // // Initialize subtask config to null
+            // ul.subtasks[subtask_key] = {
+            //     "display_name": raw_subtask["display_name"] || subtask_key,
+            //     "read_only": ("read_only" in raw_subtask) && (raw_subtask["read_only"] === true),
+            //     "inactive_opacity": 0.4
+            // };
```
Comment (TrevorBurgoyne) [inline src/index.js:730]:
```suggestion
```
remove

File: src/index.js [hunk 6]
```diff
@@ -1404,27 +720,28 @@ export class ULabel {
...
 
-            if ("inactive_opacity" in raw_subtask && typeof raw_subtask["inactive_opacity"] == "number") {
-                ul.subtasks[subtask_key]["inactive_opacity"] = Math.min(Math.max(raw_subtask["inactive_opacity"], 0.0), 1.0);
-            }
+            // if ("inactive_opacity" in raw_subtask && typeof raw_subtask["inactive_opacity"] == "number") {
+            //     ul.subtasks[subtask_key]["inactive_opacity"] = Math.min(Math.max(raw_subtask["inactive_opacity"], 0.0), 1.0);
+            // }
```
Comment (TrevorBurgoyne) [inline src/index.js:734]:
```suggestion
```
Remove

File: src/index.js [hunk 7]
```diff
@@ -1404,27 +720,28 @@ export class ULabel {
...
-            //  Initialize an empty action stream for each subtask
-            ul.subtasks[subtask_key]["actions"] = {
-                "stream": [],
-                "undone_stack": []
-            };
+            // //  Initialize an empty action stream for each subtask
+            // ul.subtasks[subtask_key]["actions"] = {
+            //     "stream": [],
+            //     "undone_stack": []
+            // };
```
Comment (TrevorBurgoyne) [inline src/index.js:744]:
```suggestion
```
Remove

File: src/index.js [hunk 8]
```diff
@@ -1526,52 +843,66 @@ export class ULabel {
...
-            };
-        }
-        else {
-            on_submit_unrolled = on_submit;
-        }
+        console.log(this)
+        console.log(toolbox_order, "Toolbox order", task_meta)
+        // // Unroll safe default arguments
+        // if (task_meta == null) { task_meta = {}; }
+        // if (annotation_meta == null) { annotation_meta = {}; }
```
Comment (TrevorBurgoyne) [inline src/index.js:866]:
```suggestion
```
Remove

File: src/index.js [hunk 9]
```diff
@@ -1526,52 +843,66 @@ export class ULabel {
...
+        // if (on_submit_unrolled.hook.constructor.name == "AsyncFunction") {
+        //     fin_on_submit_hook = on_submit_unrolled.hook;
+        // }
+        // else {
+        //     fin_on_submit_hook = async function (annotations) {
+        //         return on_submit_unrolled.hook(annotations);
+        //     };
+        // }
+
+        // this.submit_buttons1 = on_submit_unrolled
```
Comment (TrevorBurgoyne) [inline src/index.js:897]:
```suggestion
```
Remove

File: src/index.js [hunk 10]
```diff
@@ -3165,24 +2554,24 @@ export class ULabel {
...
+        // }
+        // else {
+        //     $("#" + this.config["container_id"] + " a#submit-button").attr("href", "#");
+        //     if (in_progress) {
+        //         $("#" + this.config["container_id"] + " a#submit-button").html(BUTTON_LOADER_HTML);
+        //     }
+        //     else {
+        //         $("#" + this.config["container_id"] + " a#submit-button").html(this.config["done_button"]);
+        //     }
+        // }
```
Comment (TrevorBurgoyne) [inline src/index.js:2570]:
```suggestion
```
Remove

**Review by TrevorBurgoyne — COMMENTED** [review]
Mostly suggested cleaning up some comments, imo if we want to resurrect the commented out functionality we can look back at a previous version, up to you tho. Overall the seperation into different files and the ts reworking seems very nice. It would be a lot more work but could be nice to add more detailed function helpers to those functions that still don't have them. Overall a nice chonky pr 👍

File: src/html_builder.ts (hunk 3 above)
Comment (csolbs24) [reply src/html_builder.ts]: Resolved in [c3aeb0d](https://github.com/SenteraLLC/ulabel/pull/101/commits/c3aeb0d95acf2ec0bfeb90a02b71b558a25bc8bd)

File: src/drawing_utilities.ts (hunk 2 above)
Comment (csolbs24) [reply src/drawing_utilities.ts:87]: As far as I'm aware there isn't really a good way to do this.

**Comment by csolbs24** [comment]
> It would be a lot more work but could be nice to add more detailed function helpers to those functions that still don't have them.

Yeah I think that's the plan, but for now we'll wait to do that for another refactoring passthrough. This pr is way too chonky as it is.

**Review by TrevorBurgoyne — APPROVED** [review]
LGTM 👍

File: src/annotation_operators.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import { ULabel, ULabelAnnotation, ULabelSubtask } from "..";
+
+//Given an annotation returns the confidence of that annotation
```
Comment (joshua-dean) [inline src/annotation_operators.ts:3]:
```suggestion
// Given an annotation, returns the confidence of that annotation
```

File: src/annotation_operators.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    let current_confidence = -1;
+    for (let type_of_id in annotation.classification_payloads) {
+        if (annotation.classification_payloads[type_of_id].confidence > current_confidence) {
+            current_confidence = annotation.classification_payloads[type_of_id].confidence;
+        }
+    }
+    return current_confidence;
+}
+
+//Takes in an annotation and marks it either deprecated or not deprecated.
```
Comment (joshua-dean) [inline src/annotation_operators.ts:14]:
```suggestion
// Takes an annotation and marks it as either deprecated or not deprecated.
```

File: src/annotation_operators.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+export function mark_deprecated(annotation: ULabelAnnotation, deprecated: boolean) {
+    annotation.deprecated = deprecated
+}
+
+//if the annotation confidence is less than the filter value then return true, else return false
+export function filter_low(annotation_confidence: number, filter_value: number) {
+    if (annotation_confidence < filter_value) return true
+    return false
+}
```
Comment (joshua-dean) [inline src/annotation_operators.ts:23]:
```suggestion
// Returns true if the annotation confidence is below the specified filter value.
export function filter_low(annotation_confidence: number, filter_value: number) {
    return annotation_confidence < filter_value
}
```

File: src/configuration.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        [AllowedToolboxItem.ZoomPan, ZoomPanToolboxItem],
+        [AllowedToolboxItem.AnnotationResize, AnnotationResizeItem],
+        [AllowedToolboxItem.AnnotationID, AnnotationIDToolboxItem],
+        [AllowedToolboxItem.RecolorActive, RecolorActiveItem],
+        [AllowedToolboxItem.ClassCounter, ClassCounterToolboxItem],
+        [AllowedToolboxItem.KeypointSlider, KeypointSliderItem],
+        [AllowedToolboxItem.SubmitButtons, SubmitButtons]
+    ]);
+    
+    //Change the order of the toolbox items here to change the order they show up in the toolbox
```
Comment (joshua-dean) [inline src/configuration.ts:27]:
```suggestion
    // Change the order of the toolbox items here to change the order they show up in the toolbox
```

File: src/drawing_utilities.ts [hunk 15]
```diff
@@ -0,0 +1,148 @@
+/* 
+This file is designed to hold all the functions 
+required for drawing to the canvas
+*/
+
+import { ULabelAnnotation } from './annotation';
+
+/*recieves a base color and applies a gradient to it based on its confidence
```
Comment (joshua-dean) [inline src/drawing_utilities.ts:8]:
```suggestion
/* Receives a base color and applies a gradient to it based on its confidence
```

File: src/drawing_utilities.ts [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+The reason for passing in the get confidence function is so that we can apply
+a gradient based on diffrent confidence statistics if we choose to do so*/
```
Comment (joshua-dean) [inline src/drawing_utilities.ts:10]:
```suggestion
a gradient based on diffrent confidence statistics if we choose to do so. */
```

File: src/drawing_utilities.ts [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+
+    //If the hex value is a single digit pad the front with a 0 to 
+    //ensure its two digits long
+    if (new_r_hex.length == 1) {
+        new_r_hex = "0" + new_r.toString(16)
+    }
+    if (new_g_hex.length == 1) {
+        new_g_hex = "0" + new_g.toString(16)
+    }
+    if (new_b_hex.length == 1) {
```
Comment (joshua-dean) [inline src/drawing_utilities.ts:70]: could probably use a helper as the same 5 things are done to each color channel

File: src/drawing_utilities.ts [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+/*recieves a base color and applies a gradient to it based on its confidence
+The reason for passing in the get confidence function is so that we can apply
+a gradient based on diffrent confidence statistics if we choose to do so*/
+export function apply_gradient(
+    annotation_object: ULabelAnnotation, 
+    base_color: string, 
+    get_annotation_confidence: Function, 
+    gradient_maximum: number) {
+        
+    //if the gradient toggle is checked off, then don't apply a gradient
```
Comment (joshua-dean) [inline src/drawing_utilities.ts:17]: it's probably good practice to have a space after the `//` starting a comment and to capitalize, but I it doesn't have to be done on everything for this PR, seeing as a linter can do it automatically down the road

File: src/drawing_utilities.ts (hunk 2 above)
Comment (joshua-dean) [reply src/drawing_utilities.ts:87]: we should probably pull them into their own file, or at least `blob.js`

File: src/html_builder.ts [hunk 19]
```diff
@@ -0,0 +1,499 @@
...
+                    </a><!--
```
Comment (joshua-dean) [inline src/html_builder.ts:430]: probably also a good condidate for getting moved to `blob.js` or some other constants file
not necessarily a must for this PR though, anything unresolved I can open issue trackers for

File: src/index.js [hunk 20]
```diff
@@ -2,30 +2,26 @@
 Uncertain Labeling Tool
```
Comment (joshua-dean) [inline src/index.js:2]: you ratio'd this file hardcore
nice

**Review by joshua-dean — COMMENTED** [review]
A few semantic changes suggested, but otherwise this looks pretty slick overall

File: src/html_builder.ts (hunk 19 above)
Comment (csolbs24) [reply src/html_builder.ts:430]: Do you mean like what I did in [c093011](https://github.com/SenteraLLC/ulabel/pull/101/commits/c093011b4649d1da2020053c91b005c3ba9332a9)?

File: src/drawing_utilities.ts (hunk 18 above)
Comment (csolbs24) [reply src/drawing_utilities.ts:18]: Okay, I'll keep it in mind, but I won't worry about it for this pr

File: src/html_builder.ts (hunk 19 above)
Comment (joshua-dean) [reply src/html_builder.ts:430]: yup

**Review by joshua-dean — APPROVED** [review]
LGTM 🚀

**Pull request merged**